### ISSUE TITLE
Potential Misalignment in decoding_indices Generation

### ISSUE THREAD

**Issue by Lueci4er** [Linking-ai/SCOPE#6]
Thank you for your novel and excellent work!
I have encountered some confusion when reproducing the results:

In `kv_util.py,` lines 392–404, the code uses the `h2o-fixed` metric to generate indices for gathering compressed KV cache states. Here is the relevant snippet:
```
                # prefill cache
                prefill_indices = torch.tensor(range(self.max_capacity_prompt), dtype=torch.int64).to(key_states.device)
                prefill_indices = prefill_indices.unsqueeze(0).unsqueeze(0).unsqueeze(-1).repeat(bsz, num_heads, 1, head_dim)

                # decoding cache
                decoding_indices = attn_cache[:, :, self.max_capacity_prompt:].topk(decoding_window_size - window_size, dim=-1).indices
                decoding_indices = decoding_indices.unsqueeze(-1).expand(-1, -1, -1, head_dim)
                
                # all cache
                indices = torch.cat((prefill_indices, decoding_indices), dim=2)

                k_past_compress = key_states[:, :, :-window_size, :].gather(dim = 2, index = indices)
                v_past_compress = value_states[:, :, :-window_size, :].gather(dim = 2, index = indices)
```

Specifically, the decoding_indices are generated using `attn_cache[:, :, self.max_capacity_prompt:].topk(...)`, which extracts the top-k indices from a sliced portion of `attn_cache.` However, these indices are local to the sliced region, starting from the first decoding token as index 0. When the `.gather()` operation operates on the entire tensor, it includes both the prefill and decoding regions. Since the decoding_indices do not account for the offset introduced by the prefill region, they may incorrectly reference positions in the full key_states. 

To ensure proper alignment, would it be necessary to adjust the `decoding_indices` by adding the length of the prefill region (self.max_capacity_prompt) before concatenation?

**Comment by Linking-ai** [comment]
> Thank you for your novel and excellent work! I have encountered some confusion when reproducing the results:
> 
> In `kv_util.py,` lines 392–404, the code uses the `h2o-fixed` metric to generate indices for gathering compressed KV cache states. Here is the relevant snippet:
> 
> ```
>                 # prefill cache
>                 prefill_indices = torch.tensor(range(self.max_capacity_prompt), dtype=torch.int64).to(key_states.device)
>                 prefill_indices = prefill_indices.unsqueeze(0).unsqueeze(0).unsqueeze(-1).repeat(bsz, num_heads, 1, head_dim)
> 
>                 # decoding cache
>                 decoding_indices = attn_cache[:, :, self.max_capacity_prompt:].topk(decoding_window_size - window_size, dim=-1).indices
>                 decoding_indices = decoding_indices.unsqueeze(-1).expand(-1, -1, -1, head_dim)
>                 
>                 # all cache
>                 indices = torch.cat((prefill_indices, decoding_indices), dim=2)
> 
>                 k_past_compress = key_states[:, :, :-window_size, :].gather(dim = 2, index = indices)
>                 v_past_compress = value_states[:, :, :-window_size, :].gather(dim = 2, index = indices)
> ```
> 
> Specifically, the decoding_indices are generated using `attn_cache[:, :, self.max_capacity_prompt:].topk(...)`, which extracts the top-k indices from a sliced portion of `attn_cache.` However, these indices are local to the sliced region, starting from the first decoding token as index 0. When the `.gather()` operation operates on the entire tensor, it includes both the prefill and decoding regions. Since the decoding_indices do not account for the offset introduced by the prefill region, they may incorrectly reference positions in the full key_states.
> 
> To ensure proper alignment, would it be necessary to adjust the `decoding_indices` by adding the length of the prefill region (self.max_capacity_prompt) before concatenation?


Thank you so much for your insightful and careful review of our code! You are absolutely right — there is indeed an indexing issue in the portion of `kv_util.py` you pointed out (lines 392–404).

Specifically, the `decoding_indices` are computed from a sliced segment of `attn_cache`, which yields local indices. These local indices do not account for the offset introduced by the prefill region (`self.max_capacity_prompt`). As you correctly noted, to properly align the gathered positions from the full key_states, the indices should be adjusted as follows:


```
decoding_indices = attn_cache[:, :, self.max_capacity_prompt:].topk(decoding_window_size - window_size, dim=-1).indices
decoding_indices += self.max_capacity_prompt
decoding_indices = decoding_indices.unsqueeze(-1).expand(-1, -1, -1, head_dim)
```

We sincerely appreciate your attention to detail and will correct the code and paper.

As a token of our gratitude, we will make sure to acknowledge your contribution in the paper and/or repository acknowledgments section. 

Thank you again for helping improve the correctness and quality of the project！

**Comment by callanwu** [comment]
@Lueci4er Thank you very much. We will fix and correct the results as soon as possible.

**Comment by Lueci4er** [comment]
Thank you for the quick response and for addressing the issue! 

I’m glad the feedback was helpful. If you need any further assistance while making the changes, feel free to let me know.

**Comment by Lueci4er** [comment]
Furthermore, I suggest using `.sort().values` after `.topk().indices.` The indices returned by `.topk()` are sorted by scores, which may disrupt the original index order when directly used with `.gather().` This could lead to misaligned `key_states` and `value_states,` potentially affecting subsequent inference steps.

**Comment by Linking-ai** [comment]
> Furthermore, I suggest using `.sort().values` after `.topk().indices.` The indices returned by `.topk()` are sorted by scores, which may disrupt the original index order when directly used with `.gather().` This could lead to misaligned `key_states` and `value_states,` potentially affecting subsequent inference steps.

Thank you once again for your incredibly detailed and thoughtful feedback.

We are currently testing the effects of adding:
```
decoding_indices += self.max_capacity_prompt
```
to correct the indexing alignment, and indeed, we are already observing more consistent and promising performance from our method after this fix, including an improvement of over 10 percentage points in some evaluation metrics.

We also plan to evaluate the impact of your second suggestion:
```
decoding_indices, _ = decoding_indices.sort(dim=-1)
```
to ensure the gathered `key_states` and `value_states` preserve the original temporal order, which we agree could lead to more stable and reliable results.

We truly appreciate your contributions — please feel free to continue reaching out if you notice anything else. We're very happy to keep the discussion going!!!

Once we complete a full round of testing, we will update the code accordingly.

**Comment by Lueci4er** [comment]
Thank you so much for your thoughtful response and for taking the time to thoroughly evaluate my suggestions. 

We truly appreciate the detailed explanation of the changes you are testing, as well as the positive impact they seem to be having on your method's performance. It's great to hear that the fix has already led to such significant improvements! 

Thanks again for your kind words and for being so open to collaboration.

**Comment by Linking-ai** [comment]
> > Furthermore, I suggest using `.sort().values` after `.topk().indices.` The indices returned by `.topk()` are sorted by scores, which may disrupt the original index order when directly used with `.gather().` This could lead to misaligned `key_states` and `value_states,` potentially affecting subsequent inference steps.
> 
> Thank you once again for your incredibly detailed and thoughtful feedback.
> 
> We are currently testing the effects of adding:
> 
> ```
> decoding_indices += self.max_capacity_prompt
> ```
> 
> to correct the indexing alignment, and indeed, we are already observing more consistent and promising performance from our method after this fix, including an improvement of over 10 percentage points in some evaluation metrics.
> 
> We also plan to evaluate the impact of your second suggestion:
> 
> ```
> decoding_indices, _ = decoding_indices.sort(dim=-1)
> ```
> 
> to ensure the gathered `key_states` and `value_states` preserve the original temporal order, which we agree could lead to more stable and reliable results.
> 
> We truly appreciate your contributions — please feel free to continue reaching out if you notice anything else. We're very happy to keep the discussion going!!!
> 
> Once we complete a full round of testing, we will update the code accordingly.

Thank you again for your great suggestions.

After further analysis, we believe that adding `.sort()` is not strictly necessary for correctness. Specifically:

`Causal Mask` ensures that each query can only attend to previous keys, so even if the selected keys are not in chronological order, no future information is leaked.

`RoPE` encodes absolute positional information directly into both `q` and `k`. Therefore, even if the selected keys are out of order, the inner product between queries and keys still correctly incorporates their original positional encoding.

That being said, sorting could still help improve stability and performance by preserving temporal consistency, and we plan to further experiment with it in the future.

**Comment by Lueci4er** [comment]
Thank you for your detailed explanation and thoughtful analysis. We appreciate the clarity with which you’ve outlined the reasoning behind the current implementation.
